File: app/src/main/java/com/pengxh/autodingding/BaseApplication.java
package com.pengxh.autodingding;

import android.app.Application;
import android.database.sqlite.SQLiteDatabase;
import android.util.Log;

import com.pengxh.app.multilib.utils.SaveKeyValues;
import com.pengxh.app.multilib.widget.EasyToast;
import com.pengxh.autodingding.greendao.DaoMaster;
import com.pengxh.autodingding.greendao.DaoSession;
import com.pengxh.autodingding.ui.MainActivity;
import com.pengxh.autodingding.utils.Utils;
import com.tencent.bugly.Bugly;
import com.tencent.bugly.beta.Beta;
import com.tencent.bugly.crashreport.CrashReport;

import java.util.HashSet;

import cn.jpush.android.api.JPushInterface;


public class BaseApplication extends Application {

    private static DaoSession daoSession;
    private volatile static BaseApplication application;

    @Override
    public void onCreate() {
        super.onCreate();
        application = this;
        Utils.init(this);
        EasyToast.init(this);
        SaveKeyValues.initSharedPreferences(this);
        initDataBase();
        JPushInterface.setDebugMode(true);
        JPushInterface.init(this);
        HashSet<String> strings = new HashSet<>();
        strings.add("main");
        JPushInterface.setTags(this, 0, strings);
        Log.d("push", JPushInterface.getRegistrationID(this));

        //Bugly初使化
        Bugly.init(this, "84a0dd7960", BuildConfig.DEBUG);
        Beta.autoCheckUpgrade = true;
        Beta.autoDownloadOnWifi = true;
        Beta.largeIconId = R.mipmap.ic_launcher;
        Beta.enableNotification = true;
        Beta.canShowUpgradeActs.add(MainActivity.class);

    }

    /**
     * 双重锁单例
     */
    public static BaseApplication getInstance() {
        if (application == null) {
            synchronized (BaseApplication.class) {
                if (application == null) {
                    application = new BaseApplication();
                }
            }
        }
        return application;
    }

    private void initDataBase() {
        DaoMaster.DevOpenHelper helper = new DaoMaster.DevOpenHelper(this, "DingRecord.db");
        SQLiteDatabase db = helper.getWritableDatabase();
        DaoMaster daoMaster = new DaoMaster(db);
        daoSession = daoMaster.newSession();
    }

    public static DaoSession getDaoSession() {
        return daoSession;
    }
}

File: app/src/main/java/com/pengxh/autodingding/BaseApplication.kt
package com.pengxh.autodingding

import android.app.Application
import android.content.Context
import cn.jpush.android.api.JPushInterface
import cn.vove7.andro_accessibility_api.AccessibilityApi
import com.blankj.utilcode.util.CacheDiskUtils
import com.pengxh.autodingding.greendao.DaoMaster
import com.pengxh.autodingding.greendao.DaoMaster.DevOpenHelper
import com.pengxh.autodingding.greendao.DaoSession
import com.pengxh.autodingding.service.BaseAccessibilityService
import com.pengxh.autodingding.service.GestureService
import com.pengxh.autodingding.utils.Utils

class BaseApplication : Application() {
    override fun onCreate() {
        super.onCreate()
        application = this
        Utils.init(this)
        initDataBase()
        JPushInterface.setDebugMode(true)
        JPushInterface.init(this)
        val strings = HashSet<String>()
        strings.add("main")
        JPushInterface.setTags(this, 0, strings)
        val pushRegId = JPushInterface.getRegistrationID(this)
        CacheDiskUtils.getInstance().put("pushRegId", pushRegId)

        //辅助服务
        AccessibilityApi.apply {
            BASE_SERVICE_CLS = BaseAccessibilityService::class.java
            GESTURE_SERVICE_CLS = GestureService::class.java
        }
    }

    override fun attachBaseContext(base: Context?) {
        super.attachBaseContext(base)
        application = this;
    }

    private fun initDataBase() {
        val helper = DevOpenHelper(this, "DingRecord.db")
        val db = helper.writableDatabase
        val daoMaster = DaoMaster(db)
        daoSession = daoMaster.newSession()
    }

    companion object {
        @JvmStatic
        var daoSession: DaoSession? = null
            private set

        @Volatile
        var application: BaseApplication? = null
    }
}